File: prolog/map/on_arrival.pl
:- ensure_loaded([
    '../map/paths',
    '../rules/moving',
    '../rules/inventory'
]).
% :- discontiguous path/2.

%%%main_room
on_arrival(main_room):-
    alarm_rings,
    write("    I'm in a small dark room. I'm handcuffed! I don't remeber anything. Surroundig is dirty and the room is very cold...
    An alarm clock is ringing very loudly and is exploading my head. I have to go to the desk and examine it.
    Maybe I can disable the alarm."),!,nl.

on_arrival(main_room) :-
    write("    I'm in a small dark room. It' full of creepy energy all around...
    Surroundig is dirty and the room is very cold..."),nl.
on_arrival(desk) :-
    alarm_rings,
    write("    There's the alarm. I have to examine the desk, maybe I can disable the alarm."),!,nl.
on_arrival(desk) :-
    write("    An old desk covered with dust. 
    The chair was mooved to the other end of a room, so I can't sit here..."),nl.
on_arrival(fire_place) :-
    write("    It's and big antique fireplace.
    I wish it was lit so I could warm myself up. 
    Unfortunetly without matches even the half bured CANDLE sitting on top will be no use for me..."),nl.
on_arrival(painting) :-
    write("    It's a mastepice I've seen in history books.
    It is beatifull, at least it was... before someone dameged it"),nl.
on_arrival(torn_corner) :-
    write("    It's a torn corner of a painting. 
    Fabric is deattached from glded wooden frame."),nl.
on_arrival(chair) :-
    write("    Old chair lined with red fabric. 
    Looks comfy apart of clay mud wiped all over it..."),nl.
on_arrival(vent) :-
    write("    There is a RAT in here!
    ---NEW PLACE UNLOCKED---"),nl,
    assert(subplace(vent, rat)).
on_arrival(bed) :-
    write("    In diferent situation, I would love to take a nap. Bed looks komfy, especially with that big PILLOW.
    I guess I can take a sit for a while to think..."), nl,
    spawn_item(pillow, bed),
    fail.
on_arrival(bed) :-
    !, nl.
on_arrival(sit_on_bed) :-
    write("    The bed is sqeeqing and the blanket is itchy.
    But it feels good to rest for a second."),nl.
on_arrival(window) :-
    write("    Dirty window looking out over moody forest. 
    Thermometer is hanging on the other side.
    On a window still sits a RAVEN in his NEST.
    I don't feel like opening a window with him (or her) on the other side.
    ---NEW PLACE UNLOCKED---"),nl,
    assert(subplace(window, raven)).
%%%%%%%%%%%%%%%%%%%%%%%%office
on_arrival(office) :-
    write("    I'm standing in a middle of poorly lit, destroyed office room.
    Not many things are where they supposed to be. I feel like somebody must have pluderd it long time ago..."),nl.
on_arrival(coffee_table) :-
    write("    I don't feel right about that coffee table. It's wiped clean of dust comared to the rest of the room."),nl.
on_arrival(computer) :-
    write("    Old PC. It's turned off. I hope it's OS was made in this centry..."),nl.
on_arrival(attic_entrance) :-
    write("    Old trapdoor. I can't reach it in any way possible..."),nl.
on_arrival(cupboard) :-
    write("    Wide cupboard. It's rotting away abaut to fall appart.
    It has 3 compartments. I'should check if there is something usefull in there.
    ---NEW PLACES UNLOCKED---"),
    assert(subplace(cupboard, left_compartment)),
    assert(subplace(cupboard, middle_compartment)),
    assert(subplace(cupboard, right_compartment)),nl.%has journal.

on_arrival(journal) :-
    write("    I opened an old journal I've found. It's dusty but I still want to find out more about it."),nl. %examine for more info


on_arrival(office_safe) :-
    write("    Small but sturdy. Too havy to lift. 
    It opens by unlocking a turning code mechanism"),nl.

%%%%%%%%%%%corridor
on_arrival(corridor) :-
    write("    Long corrior with paint flaking off the walls due to dense and hiumid air around.
    It`s connecting the office with the main room where I was trapped before... "),nl.
on_arrival(display_case) :-
    write("    Glass display case, there is something inside"), nl,!.
    %spawn_item(uv_flashlight,display_case), fail.
on_arrival(display_case) :-
    !, nl.
on_arrival(key_case) :-
    write("    Locked key box. Big beautifull key is stored inside. 
    It has to be opened by inputing a code on a keypad"),nl.
on_arrival(baricated_door) :-
    write("     I'm on the other side of that heavy door. It's blocked with rumble. I can unlock it now and go back to main room
    ---PATH UNLOCKED---"),
    assert(path(corridor, main_room)),
    retract(subplace(main_room, heavy_door)),
    retract(subplace(corridor,baricated_door)),
    look.
on_arrival(exit_door) :-
    write("    It's my way out. I just need a key to open it."),!,nl.

on_arrival(the_end) :-
    write("     CONGRATULATIONS YOU COMPLETED THE GAME!"),nl,
    retract(path(corridor, the_end)),!,nl.


on_arrival(_) :-
    write("     I can't see anything!").

File: haskell/src/MapEventHandler/OnArrival.hs
module MapEventHandler.OnArrival where

import Object.Place (Place (..))
import Object.Item (Item (..))
import State.GameState (GameState (alarmRings))
import Feature.Subplace (spawnSubplace, despawnSubplace)
import Feature.ItemAt (spawnItem)
import Feature.Path (spawnPath, despawnPath)
import Feature.Look (printPossibilities)

onArrival :: Place -> GameState -> IO GameState

onArrival MainRoom state = do
    if alarmRings state then do
        putStrLn "\tI'm in a small dark room. I'm handcuffed! I don't remember anything. Surrounding is dirty and the room is very cold...\n\
        \\tAn alarm clock is ringing very loudly and is exploding my head. I have to go to the desk and examine it.\n\
        \\tMaybe I can disable the alarm."
        return state
    else do
        putStrLn "\tI'm in a small dark room. It' full of creepy energy all around...\n\
        \\tSurrounding is dirty and the room is very cold..."
        return state

onArrival Desk state = do
    if alarmRings state then do
        putStrLn "\tThere's the alarm. I have to examine the desk, maybe I can disable the alarm."
        return state
    else do
        putStrLn "\tAn old desk covered with dust.\n\
        \\tThe chair was moved to the other end of a room, so I can't sit here..."
        return state

onArrival Fireplace state = do
    putStrLn "\tIt's and big antique fireplace.\n\
    \\tI wish it was lit so I could warm myself up.\n\
    \\tUnfortunately without matches even the half burned CANDLE sitting on top will be no use for me..."
    return state

onArrival Painting state = do
    putStrLn "\tIt's a masterpiece. I've seen it in history books.\n\
    \\tIt is beautiful, at least it was... before someone damaged it"
    return state

onArrival TornCorner state = do
    putStrLn "\tIt's a torn corner of a painting.\n\
    \\tFabric is detached from the gilded wooden frame."
    return state

onArrival Chair state = do
    putStrLn "\tOld chair lined with red fabric.\n\
    \\tLooks comfy apart from the clay mud wiped all over it..."
    return state

onArrival Vent state = do
    putStrLn "\tThere is a RAT in here!"
    spawnSubplace Vent Rat state

onArrival Bed state = do
    putStrLn "\tIn at different situation, I would love to take a nap. Bed looks comfy, especially with that big PILLOW.\n\
    \\tI guess I can take a sit for a while to think..."
    newState <- spawnItem Pillow Bed state
    putStrLn "\tThe bed is squeaky and the blanket is itchy.\n\
    \\tBut it feels good to rest for a second."
    return newState

onArrival Window state = do
    putStrLn "\tDirty window looking out over moody forest.\n\
    \\tThermometer is hanging on the other side.\n\
    \\tOn a window still sits a RAVEN in his NEST.\n\
    \\tI don't feel like opening a window with him (or her) on the other side."
    spawnSubplace Window Raven state

-- office

onArrival Office state = do
    putStrLn "\tI'm standing in a middle of poorly lit, destroyed office room.\n\
    \\tNot many things are where they supposed to be. I feel like somebody must have plundered it long time ago..."
    return state

onArrival CoffeTable state = do
    putStrLn "\tI don't feel right about that coffee table. It's wiped clean of dust compared to the rest of the room."
    return state

onArrival Computer state = do
    putStrLn "\tOld PC. It's turned off. I hope it's OS was made in this century..."
    return state

onArrival AtticEntrance state = do
    putStrLn "\tOld trapdoor. I can't reach it in any way possible..."
    return state

onArrival Cupboard state = do
    putStrLn "\tWide cupboard. It's rotting away about to fall apart.\n\
    \\tIt has 3 compartments. I'should check if there is something useful in there."
    return state

onArrival Journal state = do
    putStrLn "\tI opened an old journal I've found. It's dusty but I still want to find out more about it."
    return state

onArrival OfficeSafe state = do
    putStrLn "\tSmall but sturdy. Too heavy to lift.\n\
    \\tIt opens by unlocking a turning code mechanism"
    return state

-- corridor

onArrival Corridor state = do
    putStrLn "\tLong corridor with paint flaking off the walls due to dense and humid air around.\n\
    \\tIt`s connecting the office with the main room where I was trapped before... "
    return state

onArrival DisplayCase state = do
    putStrLn "\tGlass display case, there is something inside"
    return state

onArrival KeyCase state = do
    putStrLn "\tLocked key box. Big beautiful key is stored inside.\n\
    \\tIt has to be opened by inputting a code on a keypad"
    return state

onArrival BarricadedDoor state = do
    putStrLn "\tI'm on the other side of that heavy door. It's blocked with rumble. I can unlock it now and go back to main room"
    state1 <- spawnPath (Corridor, MainRoom) state
    let state2 =  despawnSubplace MainRoom HeavyDoor state1
    let state3 = despawnSubplace Corridor BarricadedDoor state2
    printPossibilities state3
    return state3

onArrival ExitDoor state = do
    putStrLn "\tIt's my way out. I just need a key to open it."
    return state

onArrival TheEnd state = do
    putStrLn "\tCONGRATULATIONS YOU COMPLETED THE GAME!"
    return $ despawnPath (Corridor, TheEnd) state

onArrival _ state = do
    putStrLn "You have arrived at a new place!"
    return state